test/threadsafe_function/threadsafe_function.cc

```diff
@@ -0,0 +1,185 @@
+    }
+
+    if (uv_thread_create(&uvThreads[1], SecondaryThread, tsFunction) != 0) {
+      Error::Fatal("DataSourceThread", "Failed to start secondary thread");
+    }
+  }
+
+  bool queueWasFull = false;
+  bool queueWasClosing = false;
+  for (int index = ARRAY_LENGTH - 1; index > -1 && !queueWasClosing; index--) {
+    napi_status status = napi_generic_failure;
+    auto callback = [](Env env, Function jsCallback, int* data) {
+      jsCallback.Call({ Number::New(env, *data) });
+    };
+
+    switch (info->type) {
+      case ThreadSafeFunctionInfo::DEFAULT:
+        status = tsFunction->BlockingCall();
+        break;
+      case ThreadSafeFunctionInfo::BLOCKING:
+        status = tsFunction->BlockingCall(&ints[index], callback);
+        break;
+      case ThreadSafeFunctionInfo::NON_BLOCKING:
+        status = tsFunction->NonBlockingCall(&ints[index], callback);
+        break;
+    }
+
+    if (info->maxQueueSize == 0) {
+      // Let's make this thread really busy for 200 ms to give the main thread a
+      // chance to abort.
+      uint64_t start = uv_hrtime();
+      for (; uv_hrtime() - start < 200000000;);
+    }
+
+    switch (status) {
+    case napi_queue_full:
+      queueWasFull = true;
+      index++;
+      // fall through
+
+    case napi_ok:
+      continue;
+
+    case napi_closing:
+      queueWasClosing = true;
+      break;
+
+    default:
+      Error::Fatal("DataSourceThread", "ThreadSafeFunction.*Call() failed");
+    }
+  }
+
+  if (info->type == ThreadSafeFunctionInfo::NON_BLOCKING && !queueWasFull) {
+    Error::Fatal("DataSourceThread", "Queue was never full");
+  }
+
+  if (info->abort && !queueWasClosing) {
+    Error::Fatal("DataSourceThread", "Queue was never closing");
+  }
+
+  if (!queueWasClosing && tsFunction->Release() != napi_ok) {
+    Error::Fatal("DataSourceThread", "ThreadSafeFunction.Release() failed");
+  }
+}
+
+static Value StopThread(const CallbackInfo& info) {
+  tsfnInfo.jsFinalizeCallback = Napi::Persistent(info[0].As<Function>());
+  bool abort = info[1].As<Boolean>();
+  if (abort) {
+    tsfn.Abort();
+  } else {
+    tsfn.Release();
+  }
+  return Value();
+}
+
+// Join the thread and inform JS that we're done.
+static void JoinTheThreads(Env /* env */,
+                           uv_thread_t* theThreads,
+                           ThreadSafeFunctionInfo* info) {
+  uv_thread_join(&theThreads[0]);
+  if (info->startSecondary) {
+    uv_thread_join(&theThreads[1]);
+  }
+
+  info->jsFinalizeCallback.Call({});
+  info->jsFinalizeCallback.Reset();
+}
+
+static Value StartThreadInternal(const CallbackInfo& info,
+    ThreadSafeFunctionInfo::CallType type) {
+  tsfnInfo.type = type;
+  tsfnInfo.abort = info[1].As<Boolean>();
+  tsfnInfo.startSecondary = info[2].As<Boolean>();
+  tsfnInfo.maxQueueSize = info[3].As<Number>().Uint32Value();
+
+  tsfn = ThreadSafeFunction::New(info.Env(), info[0].As<Function>(),
+      "Test", tsfnInfo.maxQueueSize, 2, &tsfnInfo, JoinTheThreads, uvThreads);
```

Comment:
Hi @romandev ,

I think based off of @gabrielschulhof 's comments, it is not feasible for the object itself to keep track of whether it needs to be aborted. The underlying napi tsfn is sort of "refcounted" itself through the use of `Acquire`, `Release` and -- unfortunately for ease of this wrapping implementation -- `Abort`. 

So, yeah, you are correct.. In the above scenario, the first tsfn will always exist and never die. 

My thought is that, since the tsfn is refcounted, I would have the `ThreadSafeFunction`s move and assignment operators similar to `Reference`'s, where it just sets the its instance members accordingly, and not necessarily worrying about `Abort()`ing. There is no napi "how many threads are using the tsfn" call unfortunately, so we cannot be certain whether we need to do anything to the `ThreadSafeFunction` when it is moved/assigned. Unless you implement the counter yourself, which I don't think is right at all... 

Just spitballing here... Could we do something where a `ThreadSafeFunction` object can only hold one valid `_tsFunctionValue` at a time? So in the assignment operator, you would throw an error if `this->_tsFunctionValue` is non-nullptr? Is that valid...?

Do you have thoughts @gabrielschulhof ?

Reply:
Hey @romandev ,

We talked about this in the NAPI meeting today See [here](https://www.youtube.com/watch?v=jmokE8uvc7Y&feature=youtu.be&t=2942) for the full transcript. So...

The node-addon-api wrapper itself should really be a convenience wrapper, and since the wrapper object cannot know the lifespan, it should not store instance members on itself. For example, if you have a ThreadSafeFunction that you Release() to the point where it is no longer valid (and the engine has cleaned it up), but you could still have valid `ThreadSafeFunction` object around that now has an invalid `_tsFunctionValue` reference. 

However...  It was brought up that napi tsfn does have the `Finalize` callback. So, the call to `TSFN::New()` is done in the main thread as well as the `Finalize()` callback, and these are the only places that "state" can be checked and modified (so my understanding would be the `_tsFunctionValue` member). I also remember there being an `_env` but I don't think this is necessary?)

Does this help any?

Reply:
@KevinEady,

Thank you for your input!
My new commit includes the following things:
  - Remove _env member from TSFN
  - Reset internal _tsfn pointer in Finalize callback
  - Add a check whether the internal _tsfn pointer is valid in move assign operator.

It works well in my local env.  Could you please review it again?

Reply:
@romandev I also think that, as @KevinEady mentioned above, it's acceptable to throw if one attempts to overwrite a non-empty `ThreadSafeFunction`. In fact, it's probably a good idea, because it'll help users identify leaks.